src/lib/sources.js

```diff
@@ -153,53 +154,63 @@ class Sources extends EventEmitter {
    * @return {Object}
    */
   health() {
-    const object = {
+    const obj = {
       code: STATUS_CODES.OK,
-      status: 'OK'
+      status: 'OK',
+      indices: this.indices.map((i) => i.status()),
+      sources: this.properties.sources.map((s) => s.status())
     };
 
-    const source_state = {
-      source: [],
-      count: 0
-    };
+    // Health logic:
+    // Indices:
+    // 500 - any index is !ok
+    // 503 - any index is !ready
+    // 200 - all indices are both ok and ready
+    // Sources:
+    // 500 - all sources are !ok
+    // 503 - any source is !ready
+    // 200 - all sources both ok and ready
+    //
+    const notReady = (s) => !s.ready;
+    const unhealthy = (s) => !s.ok;
+
+    if (this.indices.some(unhealthy)) {
+      const u = this.indices.find((i) => !i.ok);
+
+      obj.code = STATUS_CODES.INTERNAL_SERVER_ERROR;
+      obj.status = u.status().state;
+
+      return obj;
+    }
 
-    object.indices = this.indices.map((source) => {
-      // TODO This logic is fairly ham-fisted right now. It'll work because nothing
-      // is setting `state` to WARNING at the moment. When we do start supporting a
-      // WARNING state, this will have to become aware that OK < WARING < ERROR.
-      if (!source.ready) {
-        object.code = STATUS_CODES.SERVICE_UNAVAILABLE;
-        object.status = source.state;
-      }
+    if (this.indices.some(notReady)) {
+      const u = this.indices.find((i) => !i.ready);
 
-      if (!source.ok) {
-        object.code = STATUS_CODES.INTERNAL_SERVER_ERROR;
-        object.status = source.state;
-      }
+      obj.code = STATUS_CODES.SERVICE_UNAVAILABLE;
+      obj.status = u.status().state;
 
-      return source.status();
-    });
+      return obj;
+    }
 
-    object.sources = this.properties.sources.map((source) => {
-      if (!source.ready) {
-        object.code = STATUS_CODES.SERVICE_UNAVAILABLE;
-        object.status = source.state;
-      }
+    if (this.properties.sources.every(unhealthy)) {
+      const u = this.properties.sources.find((s) => !s.ok);
 
-      if (!source.ok) {
-        source_state.source.push(source.state);
-        source_state.count = source_state.count + 1;
-      }
+      obj.code = STATUS_CODES.INTERNAL_SERVER_ERROR;
+      obj.status = u.status().state;
 
-      return source.status();
-    });
+      return obj;
+    }
+
+    if (this.properties.sources.some(notReady)) {
+      const u = this.properties.sources.find((s) => !s.ready);
+
+      obj.code = STATUS_CODES.SERVICE_UNAVAILABLE;
+      obj.status = u.status().state;
 
-    if (object.sources.length === source_state.count) {
-      object.code = STATUS_CODES.INTERNAL_SERVER_ERROR;
-      object.status = source_state.source[source_state.count - 1];
+      return obj;
     }
 
-    return object;
+    return obj;
```

Comment:
This method seems a lot neater - more JS way of doing things?
Setting the states to begin, returning once one error code gets updated?

Reply:
I'm not jazzy about the implementation but it captures our logic well. I'm definitely open to feedback about the actual code.